core/storage/question/gae_models.py

```diff
@@ -283,62 +282,68 @@ def get_question_skill_links_based_on_mastery_equidistributed_by_skill(
                 If not enough questions for a skill, just return all questions
                 it links to. The order of questions will follow the order of
                 given skill ids, and the order of questions for the same skill
-                follows the difference between skill difficulty and mastery.
+                follows the absolute value of the difference between skill
+                difficulty and the requested difficulty.
         """
-        if (degrees_of_mastery is not None and
-                not isinstance(degrees_of_mastery, dict)):
-            raise Exception('Degrees of mastery must be a dictionary.')
-
         question_count_per_skill = int(
             math.ceil(float(total_question_count) / float(len(skill_ids))))
 
         question_skill_link_models = []
         existing_question_ids = []
 
         for skill_id in skill_ids:
-            if degrees_of_mastery is None or not skill_id in degrees_of_mastery:
-                degree_of_mastery = 0.0
-            else:
-                degree_of_mastery = degrees_of_mastery[skill_id]
-                if degree_of_mastery is None:
-                    degree_of_mastery = 0.0
-
             query = cls.query(cls.skill_id == skill_id)
 
-            # Fetch QuestionSkillLinkModels with difficulty larger than
-            # mastery and sort them by increasing difficulty.
-            harder_questions_query = query.filter(
-                cls.skill_difficulty >= degree_of_mastery)
-            harder_questions_query = harder_questions_query.order(
-                cls.skill_difficulty)
+            equal_questions_query = query.filter(
+                cls.skill_difficulty == difficulty_requested)
+            # Get as many requested difficulty questions as possible.
             new_question_skill_link_models = (
-                harder_questions_query.fetch(question_count_per_skill))
-
-            # Fetch QuestionSkillLinkModels with difficulty smaller than
-            # mastery and sort them by decreasing difficulty.
-            easier_questions_query = query.filter(
-                cls.skill_difficulty < degree_of_mastery)
-            easier_questions_query = easier_questions_query.order(
-                -cls.skill_difficulty)
-            new_question_skill_link_models.extend(
-                easier_questions_query.fetch(question_count_per_skill))
-
-            # Deduplicate if the same question is linked to multiple skills.
-            # Note: This deduplication might make the number of returning
-            # questions too low, but since we fetched double size of models
-            # above, the chance should be fairly small.
+                equal_questions_query.fetch(question_count_per_skill * 2))
```

Comment:
Here and elsewhere, consider adding a comment explaining the "* 2". I think it's not obvious why we do that.

Reply:
Done.